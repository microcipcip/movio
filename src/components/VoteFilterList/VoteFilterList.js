import React from 'react'
import styled from 'styled-components'
import { FaCaretDown, FaCaretUp } from 'react-icons/fa'
import * as s from 'styles/vars'
import { TMDB_VOTE_RATING_LIST } from 'config'
import CinemaContext from 'components/App/CinemaContext'
import VoteFilter from 'components/VoteFilter'

class VoteFilterList extends React.Component {
  state = {
    isDropdownActive: false,
  }

  toggleSelect = () => {
    const { isDropdownActive } = this.state
    this.setState(() => ({ isDropdownActive: !isDropdownActive }))
  }

  closeSelectOnFilter = (setVoteFilter, val) => {
    this.toggleSelect()
    setVoteFilter(val)
  }

  render() {
    const { isDropdownActive } = this.state
    const VoteFilterIcon = isDropdownActive ? FaCaretUp : FaCaretDown
    return (
      <CinemaContext.Consumer>
        {({ setVoteFilter, voteFilter }) => (
          <VoteFilterListStyled>
            <VoteFilterSelect onClick={this.toggleSelect}>
              <VoteFilterSelectText>
                Vote average > {voteFilter}
              </VoteFilterSelectText>{' '}
              <VoteFilterIcon />
            </VoteFilterSelect>
            <VoteFilterOptions active={isDropdownActive}>
              {TMDB_VOTE_RATING_LIST.map(rating => (
                <VoteFilter
                  key={rating}
                  ratingValue={rating}
                  isActive={voteFilter === rating}
                  setVoteFilter={val =>
                    this.closeSelectOnFilter(setVoteFilter, val)
                  }
                />
              ))}
            </VoteFilterOptions>
          </VoteFilterListStyled>
        )}
      </CinemaContext.Consumer>
    )
  }
}

const VoteFilterListStyled = styled.div`
  display: flex;
  position: relative;
`

const VoteFilterSelect = styled.div`
  cursor: pointer;
  display: flex;
  justify-content: space-between;
  align-items: center;
  min-width: 200px;
  padding: 4px ${s.gutter};
  color: ${props => props.theme.voteFilterSelectTextColor};
  border: 1px solid ${props => props.theme.voteFilterSelectBorderColor};
`

const VoteFilterSelectText = styled.div`
  margin-right: 8px;
`

const VoteFilterOptions = styled.div`
  position: absolute;
  top: 100%;
  width: 100%;
  left: 0;
  padding: 2px 0;
  background-color: ${props => props.theme.mainSectionFilterBarBg};
  opacity: 0;
  ${props =>
    props.active &&
    `
    opacity: 1;
  `}
`

export default VoteFilterList
