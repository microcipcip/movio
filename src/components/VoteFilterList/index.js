import Index from './VoteFilterList'

export default Index
