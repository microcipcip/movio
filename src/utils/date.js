export const addZero = date => ('0' + date).slice(-2)

export const formatMovieDate = movieDate => {
  const d = new Date(movieDate)
  return `${addZero(d.getDate())} ${addZero(
    d.getMonth() + 1
  )} ${d.getFullYear()}`
}
