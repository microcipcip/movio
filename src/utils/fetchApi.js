// wrapper for the fetch API
// because it doesn't throw errors when response status is 400
import { TMDB_API_KEY, TMDB_API_URL, TMDB_API_LANG } from 'config/index'

const parseJSON = response =>
  new Promise(resolve =>
    response.json().then(json =>
      resolve({
        status: response.status,
        ok: response.ok,
        json,
      })
    )
  )

export default url => {
  const queryOrAmpersand = url.includes('?') ? '&' : '?'
  const query = `${TMDB_API_URL}${url}${queryOrAmpersand}api_key=${TMDB_API_KEY}&language=${TMDB_API_LANG}`

  return new Promise((resolve, reject) =>
    fetch(query)
      .then(parseJSON)
      .then(response => {
        if (response.ok) {
          return resolve(response.json)
        }
        // extract the error from the server's json
        return reject(response.json.meta.error)
      })
      .catch(err =>
        reject({
          networkError: err.message,
        })
      )
  )
}
