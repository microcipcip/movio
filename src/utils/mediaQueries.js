import { css } from 'styled-components'

// Breakpoint
export const mediaWidthRule = rule =>
  `${{ up: 'min', down: 'max' }[rule] || 'min'}-width`

export const ruleTemplate = (rule, width) => `(${rule}: ${width}px)`
export const mediaTemplate = rules => `@media only screen and ${rules}`

export const getSmallestMedia = breakpoints => {
  const keys = Object.keys(breakpoints)

  return Number(
    keys.reduce((acc, key) => {
      if (acc) {
        return acc > breakpoints[key] ? breakpoints[key] : acc
      }
      return breakpoints[key]
    }, undefined)
  )
}

export const getNextMedia = (breakpoints, width) => {
  const keys = Object.keys(breakpoints)

  const nextBreakpoint = Number(
    keys.reduce((acc, key) => {
      const current = breakpoints[key]
      if (acc && width - current <= 0) {
        return width - acc > Math.abs(width - current) || width - acc === 0
          ? current
          : acc
      }
      return getSmallestMedia(breakpoints)
    }, undefined)
  )

  if (nextBreakpoint === width) {
    throw new Error(
      `The breakpoint of value ${nextBreakpoint} is the heighst there is, did you mean to use "[media].up"?`
    )
  }
  return nextBreakpoint
}

export const mediaRules = (breakpoints, widthKey, rule, boundKey) => {
  const width = breakpoints[widthKey]
  const bound = breakpoints[boundKey]
  let baseWidthRule = mediaWidthRule(rule)
  let boudWidthRule

  let baseRule = ruleTemplate(baseWidthRule, width)
  let boundRule

  if (bound && width) {
    // Get correct rule based on width relative to bound
    baseWidthRule = mediaWidthRule(bound <= width ? 'down' : 'up')
    boudWidthRule = mediaWidthRule(bound <= width ? 'up' : 'down')

    baseRule = ruleTemplate(
      mediaWidthRule(bound <= width ? 'down' : 'up'),
      width
    )
    boundRule = ruleTemplate(boudWidthRule, bound)
    return [].concat([baseRule], bound ? [boundRule] : []).join(' and ')
  }

  if (!bound && rule === 'only') {
    // Get correct rule based on width relative to bound
    boudWidthRule = mediaWidthRule(bound <= width ? 'up' : 'down')
    boundRule = ruleTemplate(boudWidthRule, getNextMedia(breakpoints, width))
  }

  return [].concat([baseRule], boundRule ? [boundRule] : []).join(' and ')
}

export const getMixin = (breakpoints, key, rule = 'up') => bound => (
  ...args
) => css`
  ${mediaTemplate(mediaRules(breakpoints, key, rule, bound))} {
    ${css(...args)}
  }
`

export const getMediaShorthands = (breakpoints, rule) =>
  Object.keys(breakpoints).reduce(
    (acc, key) => ({
      ...acc,
      [key]: getMixin(breakpoints, key, rule)(),
    }),
    {}
  )

export const getMedia = breakpoints => {
  const mediasUp = getMediaShorthands(breakpoints, 'up')
  const list = Object.keys(breakpoints)

  return {
    ...mediasUp,
    up: widthKey => getMixin(breakpoints, widthKey, 'up')(),
    down: widthKey => getMixin(breakpoints, widthKey, 'down')(),
    only: (widthKey, boundKey) =>
      getMixin(breakpoints, widthKey, 'only')(boundKey),
    list,
  }
}

export default getMedia
