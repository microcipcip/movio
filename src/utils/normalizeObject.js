export default (acc, cur) => {
  acc[cur.id] = cur
  return acc
}
