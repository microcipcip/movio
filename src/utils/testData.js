export const genreList = {
  1: {
    id: 1,
    name: 'Action',
    isChecked: false,
  },
  2: {
    id: 2,
    name: 'Adventure',
    isChecked: false,
  },
  3: {
    id: 3,
    name: 'Animation',
    isChecked: false,
  },
  4: {
    id: 4,
    name: 'Comedy',
    isChecked: false,
  },
  5: {
    id: 5,
    name: 'Crime',
    isChecked: false,
  },
}

export const movieList = [
  {
    id: 1,
    vote_average: 6.5,
    title: 'Venom',
    popularity: 423.154,
    poster_path: '/2uNW4WbgBXL25BAbXGLnLqX71Sw.jpg',
    genre_ids: [1, 2],
    release_date: '2018-10-03',
  },
  {
    id: 2,
    vote_average: 7,
    title: 'Fantastic Beasts: The Crimes of Grindelwald',
    popularity: 220.606,
    poster_path: '/uyJgTzAsp3Za2TaPiZt2yaKYRIR.jpg',
    genre_ids: [3, 4],
    release_date: '2018-11-14',
  },
  {
    id: 3,
    vote_average: 8.2,
    title: 'Bohemian Rhapsody',
    popularity: 211.565,
    poster_path: '/lHu1wtNaczFPGFDTrjCSzeLPTKN.jpg',
    genre_ids: [5],
    release_date: '2018-10-24',
  },
  {
    id: 4,
    vote_average: 5.9,
    title: 'Mortal Engines',
    popularity: 201.549,
    poster_path: '/uXJVpPXxZO4L8Rz3IG1Y8XvZJcg.jpg',
    genre_ids: [1],
    release_date: '2018-12-06',
  },
  {
    id: 5,
    vote_average: 6,
    title: 'Robin Hood',
    popularity: 171.519,
    poster_path: '/AiRfixFcfTkNbn2A73qVJPlpkUo.jpg',
    genre_ids: [2],
    backdrop_path: '/AuA50D7T7S7OEVcGo0ZKaMhERn0.jpg',
    release_date: '2018-11-20',
  },
  {
    id: 6,
    vote_average: 5.9,
    title: 'The Seven Deadly Sins: Prisoners of the Sky',
    popularity: 150.182,
    poster_path: '/r6pPUVUKU5eIpYj4oEzidk5ZibB.jpg',
    genre_ids: [3],
    release_date: '2018-08-18',
  },
  {
    id: 7,
    vote_average: 7.5,
    title: 'Ralph Breaks the Internet',
    popularity: 141.193,
    poster_path: '/m110vLaDDOCca4hfOcS5mK5cDke.jpg',
    genre_ids: [4, 5],
    release_date: '2018-11-20',
  },
]
