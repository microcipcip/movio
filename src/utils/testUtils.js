import React from 'react'
import { render } from 'react-testing-library'
import { ThemeProvider } from 'styled-components'
import 'jest-styled-components'
import 'jest-dom/extend-expect'
import { genreList, movieList } from 'utils/testData'
import * as s from 'styles/vars'

import CinemaContext, { cinemaContextState } from 'components/App/CinemaContext'

const customRender = (node, data = {}) => {
  return render(
    <CinemaContext.Provider value={Object.assign(cinemaContextState, data)}>
      <ThemeProvider theme={s.theme.dark}>{node}</ThemeProvider>
    </CinemaContext.Provider>
  )
}

const domToString = el => {
  const clnEl = el.cloneNode(true)
  const tmp = document.createElement('div')
  tmp.appendChild(clnEl)
  return tmp.innerHTML
}

// re-export everything
export * from 'react-testing-library'
export { customRender, domToString, genreList, movieList }
