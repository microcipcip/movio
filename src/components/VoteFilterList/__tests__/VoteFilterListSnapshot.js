import React from 'react'
import { customRender, movieList } from 'utils/testUtils'
import VoteFilterList from '../VoteFilterList'

it('Should create snapshot', async () => {
  const { container } = customRender(<VoteFilterList />, { movieList })
  expect(container).toMatchSnapshot()
})
