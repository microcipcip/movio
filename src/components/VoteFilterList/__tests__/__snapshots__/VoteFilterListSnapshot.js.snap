// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create snapshot 1`] = `
.c4 {
  cursor: pointer;
  padding: 2px 10px;
}

.c4:hover {
  background-color: #161616;
}

.c5 {
  cursor: pointer;
  padding: 2px 10px;
  background-color: #161616;
}

.c5:hover {
  background-color: #161616;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 200px;
  padding: 4px 10px;
  color: #919191;
  border: 1px solid #2b2b2b;
}

.c2 {
  margin-right: 8px;
}

.c3 {
  position: absolute;
  top: 100%;
  width: 100%;
  left: 0;
  padding: 2px 0;
  background-color: #111111;
  opacity: 0;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="VoteFilterSelect"
    >
      <div
        class="c2"
      >
        Vote average &gt; 
        3
      </div>
       
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 320 512"
        width="1em"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        />
      </svg>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        0
      </div>
      <div
        class="c4"
      >
        0.5
      </div>
      <div
        class="c4"
      >
        1
      </div>
      <div
        class="c4"
      >
        1.5
      </div>
      <div
        class="c4"
      >
        2
      </div>
      <div
        class="c4"
      >
        2.5
      </div>
      <div
        class="c5"
      >
        3
      </div>
      <div
        class="c4"
      >
        3.5
      </div>
      <div
        class="c4"
      >
        4
      </div>
      <div
        class="c4"
      >
        4.5
      </div>
      <div
        class="c4"
      >
        5
      </div>
      <div
        class="c4"
      >
        5.5
      </div>
      <div
        class="c4"
      >
        6
      </div>
      <div
        class="c4"
      >
        6.5
      </div>
      <div
        class="c4"
      >
        7
      </div>
      <div
        class="c4"
      >
        7.5
      </div>
      <div
        class="c4"
      >
        8
      </div>
      <div
        class="c4"
      >
        8.5
      </div>
      <div
        class="c4"
      >
        9
      </div>
      <div
        class="c4"
      >
        9.5
      </div>
      <div
        class="c4"
      >
        10
      </div>
    </div>
  </div>
</div>
`;
